### PR TITLE
Procgen wrapper

### PR DESCRIPTION
**Pull request by cswinter**

### CODE REVIEW

File: enn_zoo/enn_zoo/procgen_env/big_fish.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    "climber_spawn_x",
+]
+
+BIG_FISH_FEATS = ["fish_eaten", "r_inc"]
+EXTRA_FEATS = ["dsize", "bigger"]
+
+
+class BigFish(Environment):
+    def __init__(self, extra_feats: bool = False) -> None:
+        self.env = ProcgenGym3Env(
```
Comment (vwxyzjn) [inline enn_zoo/enn_zoo/procgen_env/big_fish.py:47]: I suggest making the `distribution_mode` an argument. `ProcgenGym3Env` uses 
 the hard distribution mode by default: https://github.com/openai/procgen/blob/4b5e5d5806c89303a625386a4a6aae200327fdb5/procgen/env.py#L217, but I'd recommend using the `easy` mode by default instead to save compute.

File: enn_zoo/enn_zoo/procgen_env/big_fish.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            num=1,
+            env_name="bigfish",
+            start_level=0,
+            num_levels=1000000,
```
Comment (vwxyzjn) [inline enn_zoo/enn_zoo/procgen_env/big_fish.py:51]: > num_levels=0 - The number of unique levels that can be generated. Set to 0 to use unlimited levels.

Maybe set `num_levels = 0` by default?

File: configs/procgen/bigfish.ron [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    optim: (
+        bs: 8192,
+        lr: 0.01,
+    ),
+    ppo: (
+        ent_coef: 0.1,
+        anneal_entropy: true,
+        gamma: 0.99,
+    ),
+)
```
Comment (vwxyzjn) [inline configs/procgen/bigfish.ron:29]: These are the hyperparameters from the procgen paper:

```
ExperimentConfig(
    env: (
        id: "BigFish",
    ),
    rollout: (
        num_envs: 64,
        steps: 256,
        processes: 1, // we should use process = 1 becasue procgen under the hood leverages multi threads and processes?
    ),
    total_timesteps: 25000000,
    net: (
        d_model: 64,
        n_layer: 2,
        translation: (
            position_features: ["x", "y"],
            reference_entity: "Player",
            add_dist_feature: true,
        ),
    ),
    optim: (
        bs: 8192,
        lr: 0.0005,
        max_grad_norm: 0.5,
    ),
    ppo: (
        ent_coef: 0.01,
        anneal_entropy: false,
        gamma: 0.999,
    ),
) 
``` 

derived from here: https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/ppo_procgen.py, which results in very accurate reproduction:

image

I'd suggest at least matching 

```
    num_envs: 64,
    steps: 256,
    total_timesteps: 25000000,
```

**Review by vwxyzjn — COMMENTED** [review]
Looks great!!

**Comment by cswinter** [comment]
Braindump:
- current integration is quite inefficient, i'm getting ~4K SPS but majority of time is spent on env, probably we could get at least 10K-20K
  - one gym3 vectorized env is created for every environment, instead we should directly implement an entity-gym `VecEnv` and push all parallelization into procgen
  - deserializer could be a lot faster, it should never materialized any dataclasses, write directly into numpy arrays, skip all fields which we aren't using, skip tilemap unless it's used, deserializes multiple contiguous fields of the same type in one go
- at least three more environments which have only entities would be very easy to add support for: starfighter, boss fight, leaper, maybe also plunder
  - entities are identical across all environments
  - environments write out a little bit of additional info at the end which is specific to the environment (e.g. [BigFish::serialize](https://github.com/openai/procgen/blob/5e1dbf341d291eff40d1f9e0c0a0d5003643aebf/procgen/src/games/bigfish.cpp#L108)) which may contain important additional features
- other environments include a tilemap as well which doesn't map nicely to anything in entity-gym currently
  - running `python deserialize.py` will print out the tilemap
  - some of these could probably be converted into entities, e.g. coinrun tilemap below (truncated, full tilemap is padded to 64x64)
  - G is goal, W is wall, L is lava, "B" is invisible barrier used by enemy AI, when enemy collides with barrier it reverses direction.
  - could just have one entity for each horizontal position that has a feature for the height of the platform
```
WW              WWWWWWWWWWWWWWWWWWWWWW
WW              WWWWWWWWWWWWWWWWWWWWWW
WW       B     GWWWWWWWWWWWWWWWWWWWWWW
WW        WWWWWWWWWWWWWWWWWWWWWWWWWWWW
WW   B    WWWWWWWWWWWWWWWWWWWWWWWWWWWW
WW    WWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWW
WW    WWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWW
WW    WWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWW
xWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWW
```

**Comment by cswinter** [comment]
Getting good performance turned out to be a bit harder than I thought, the `get_state` method turns out to be a bottleneck. I've eliminated some inefficiencies in the parsing, which gives a nice speedup, but still far from what Procgen is capable of. There's not further easy gains though without optimizing/parallelizing the `get_state` method and Procgen serialization logic. Since `get_state` isn't multithreaded afaik, using the in-built vectorized environment and parallelization doesn't really help, and so sticking with the current design where we have one single-env vectorized env for every environment and do parallelization on our side is both simpler and more performant.

**Pull request merged**